keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,359 @@
+way to configure different runtimes in the cluster, surface their properties (both to the cluster &
+the user), and provides a mechanism for pods to select a runtime.
+
+[Sandboxes]: https://docs.google.com/document/d/1QQ5u1RBDLXWvC8K3pscTtTRThsOeBSts_imYEoRyw8A/edit
+
+### Goals
+
+- Provide a mechanism for surfacing container runtime properties to the control plane
+- Support multiple runtimes per-cluster, and provide a mechanism for users to select the desired
+  runtime
+
+### Non-Goals
+
+- RuntimeClass is NOT a general policy mechanism
+- RuntimeClass is NOT "NodeClass". Although different nodes may run different runtimes, in general
+  RuntimeClass should not be a cross product of runtime properties and node properties.
+
+The following goals are out-of-scope for the initial implementation, but may be explored in a future
+iteration:
+
+- Automatic runtime or feature discovery - initially RuntimeClasses are manually defined (by the
+  cluster admin or provider), and are asserted to be an accurate representation of the runtime.
+- Scheduling in heterogeneous clusters - it is possible to operate a heterogeneous cluster
+  (different runtime configurations on different nodes) through scheduling primitives like
+  `NodeAffinity` and `Taints+Tolerations`, but the user is responsible for setting these up and
+  automatic runtime-aware scheduling is out-of-scope.
+- Define standardized or conformant runtime classes - although I would like to declare some
+  predefined RuntimeClasses with specific properties, doing so is out-of-scope for this initial KEP.
+- [Pod Overhead][] - Although RuntimeClass is likely to be the configuration mechanism of choice,
+  the details of how pod resource overhead will be implemented is out of scope for this KEP.
+- Provide a mechanism to dynamically register or provision additional runtimes.
+- Requiring specific RuntimeClasses according to policy. This should be addressed by other
+  cluster-level policy mechanisms, such as PodSecurityPolicy.
+- "Fitting" a RuntimeClass to pod requirements - In other words, specifying runtime properties and
+  letting the system match an appropriate RuntimeClass, rather than explicitly assigning a
+  RuntimeClass by name. This approach can increase portability, but can be added seamlessly in a
+  future iteration.
+
+[Pod Overhead]: https://docs.google.com/document/d/1EJKT4gyl58-kzt2bnwkv08MIUZ6lkDpXcxkHqCvvAp4/edit
+
+### User Stories
+
+- As a cluster operator, I want to provide multiple runtime options to support a wide variety of
+  workloads. Examples include native linux containers, "sandboxed" containers, and windows
+  containers.
+- As a cluster operator, I want to provide stable rolling upgrades of nodes.
+- As an application developer, I want to select the runtime that best fits my workload.
+- As an application developer, I don't want to study the nitty-gritty details of different runtime
```

Comment:
Yet, the application developer will need to bear the cognitive load of understanding the trade-offs between runtimes since they will have to write the pod spec to select the runtime explicitly rather than expressing the attributes they require of the runtime ("Fitting" from above)

Reply:
Yes, but this shifts some of the responsibility. For example, the cluster admin can set up the `native` RuntimeClass that accurately reflects the features supported & used in the specific cluster, and tell developers to use that instead of `legacy` (or even require it with policy), that way the developer doesn't need to read the details of the specific node OS / runtime versions - they can just see the supported features directly in the RuntimeClass. If the admin sets up a `sandboxed` class, the developer can use it with a reasonable assumption of security (if they trust the admin), without needing to learn all the knobs to tune - it's a simplification. We also might add "fitting" in the future, it's not off the table.

Reply:
i can imagine runtime classes like containerd-default, containerd-secure, etc or default, secure, super-secure :-)